server/lib/scheduler/Scheduler.ts

```diff
@@ -261,126 +232,118 @@ class Scheduler {
-            userId,
-            t,
-          );
-          return isHonestOracle || isMaliciousOracle;
-        },
-      );
+      const oracleTrees = await filter(highestPriorityTrees, t => {
+        // TODO: make sure fetchTrees.. is properly getting the oracle relations array
+        return t.oracleRelations.length > 0;
+      });
 
-      let nonOracleTreesToConsider = _.difference(
-        highestPriorityTrees,
-        oracleTreesToConsider,
-      );
+      let oracleTreesToConsider = oracleTrees;
 
       while (oracleTreesToConsider.length > 0) {
-        const leastRecentlyWorkedOnTreesToConsider = await this.getTreesWorkedOnLeastRecentlyByUser(
-          userId,
-          oracleTreesToConsider,
-        );
-        const randomlySelectedTree = pickRandomItemFromArray(
-          leastRecentlyWorkedOnTreesToConsider,
-        );
-        const workspacesInTree = await this.fetchAllWorkspacesInTree(
-          randomlySelectedTree,
+        const rootWorkspacesOfTrees: Workspace[] = oracleTreesToConsider.map(
+          t => t.rootWorkspace,
         );
-
-        const isHonestOracle = await this.isUserHonestOracleForRootWorkspace(
+        const leastRecentlyWorkedOnTreeRootWorkspaces: Workspace[] = await this.getTreesWorkedOnLeastRecentlyByUser(
           userId,
-          randomlySelectedTree,
+          rootWorkspacesOfTrees,
         );
-        const isMaliciousOracle = await this.isUserMaliciousOracleForRootWorkspace(
-          userId,
-          randomlySelectedTree,
+        const randomlySelectedRootWorkspace = pickRandomItemFromArray(
+          leastRecentlyWorkedOnTreeRootWorkspaces,
         );
 
-        let oracleEligibleWorkspaces = workspacesInTree;
-        if (isHonestOracle) {
-          oracleEligibleWorkspaces = await this.filterByWhetherEligibleForHonestOracle(
-            workspacesInTree,
-          );
-        } else {
-          oracleEligibleWorkspaces = await this.filterByWhetherEligibleForMaliciousOracle(
-            workspacesInTree,
-          );
-        }
+        const selectedTree: Tree = _.find(oracleTreesToConsider, tree => {
+          return tree.rootWorkspace.id === randomlySelectedRootWorkspace.id;
+        });
 
-        oracleEligibleWorkspaces = await this.filterByStaleness(
-          userId,
-          oracleEligibleWorkspaces,
+        const isMaliciousInTree: boolean =
+          selectedTree.oracleRelations[0].isMalicious;
+
+        const workspacesInTree: Workspace[] = await this.fetchAllWorkspacesInTree(
+          randomlySelectedRootWorkspace,
         );
 
-        const workspacesToConsider = await this.filterByWhetherCurrentlyBeingWorkedOn(
+        const oracleEligibleWorkspaces: Workspace[] = isMaliciousInTree
+          ? this.filterByWhetherEligibleForMaliciousOracle(workspacesInTree)
+          : this.filterByWhetherEligibleForHonestOracle(workspacesInTree);
+
+        const staleOracleEligibleWorkspaces = this.filterByStaleness(
```

Comment:
It doesn't look like `staleOracleEligibleWorkspaces` is being used. We do need to make sure to only ever assign a stale workspace though.

Reply:
Good catch!